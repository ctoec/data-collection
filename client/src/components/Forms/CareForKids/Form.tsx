import React, { useState, useContext } from 'react';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { CareForKidsField } from './CareForKidsField';
import { EditFormProps } from '../types';
import AuthenticationContext from '../../../contexts/AuthenticationContext/AuthenticationContext';
import { Child } from '../../../shared/models';
import { apiPut } from '../../../utils/api';
/*
 * Basic functional component designed to allow user to edit
 * the Care For Kids field of a Child object.
 */
export const CareForKidsForm: React.FC<EditFormProps> = ({
  child,
  onSuccess,
}) => {
  const { accessToken } = useContext(AuthenticationContext);
  const [isSaving, setIsSaving] = useState(false);

  if (!child) return <></>;

  const onSubmit = (updatedChild: Child) => {
    setIsSaving(true);
    apiPut(`children/${child.id}`, updatedChild, { accessToken })
      .then(() => onSuccess())
      .catch((err) => {
        console.log(err);
      })
      .finally(() => setIsSaving(false));
  };

  return (
    <div className="grid-container margin-top-2">
      <h2 className="grid-row">Care 4 Kids</h2>
      <div>
        <Form<Child>
          className="CareForKidsForm"
          data={child}
          onSubmit={onSubmit}
          noValidate
          autoComplete="off"
        >
          <CareForKidsField />
          <div className="grid-row margin-top-2">
            <FormSubmitButton text={isSaving ? 'Saving...' : 'Save'} />
          </div>
        </Form>
      </div>
    </div>
  );
};
