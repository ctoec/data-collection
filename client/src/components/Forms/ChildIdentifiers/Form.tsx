import React, { useContext, useEffect, useState } from 'react';
import {
  SasidField,
  FirstNameField,
  MiddleNameField,
  LastNameField,
  SuffixField,
  DateOfBirthField,
  BirthCertificateFieldSet,
} from './Fields';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { RecordFormProps } from '../types';
import AuthenticationContext from '../../../contexts/AuthenticationContext/AuthenticationContext';
import { Child, UniqueIdType } from '../../../shared/models';
import { apiPost, apiPut } from '../../../utils/api';
import useIsMounted from '../../../hooks/useIsMounted';
import { useValidationErrors } from '../../../hooks/useValidationErrors';
import { getValidationStatusForFields } from '../../../utils/getValidationStatus';
import { useHistory } from 'react-router-dom';
import { UniqueIdField } from './Fields/UniqueId';

// The fields we use to check to see if this form has errors or missing info
const childIdentifiersFields = [
  'firstName',
  'lastName',
  'birthdate',
  'birthCertificateType',
  'birthTown',
  'birthState',
  'birthCertificateId',
];
export const doesChildIdFormHaveErrors = (child?: Child) =>
  child ? !!getValidationStatusForFields(child, childIdentifiersFields) : true;

export const ChildIdentifiersForm = ({
  child: inputChild,
  afterSaveSuccess,
  hideHeader = false,
  hideErrors,
  showFieldOrFieldset = () => true,
  setAlerts,
}: RecordFormProps) => {
  const { accessToken } = useContext(AuthenticationContext);
  const history = useHistory();
  const isMounted = useIsMounted();
  const [saving, setSaving] = useState(false);

  if (!inputChild) {
    throw new Error('Child info rendered without child');
  }
  // Must keep track of child locally in case creation fails
  const [child, updateChild] = useState(inputChild);
  useEffect(() => {
    updateChild(inputChild);
  }, [inputChild]);
  const { errorsHidden } = useValidationErrors(hideErrors);

  const onFinally = () => {
    if (isMounted()) {
      setSaving(false);
    }
  };

  const onFormSubmit = (_child: Child) => {
    setSaving(true);
    if (!child.id) {
      apiPost('children', _child, {
        accessToken,
      })
        .then((res) => {
          history.replace({ pathname: `/create-record/${res.id}` });
          afterSaveSuccess();
        })
        .catch((err) => {
          if (err.data) {
            updateChild(err.data);
          }
          console.error(err);
          setAlerts([
            { type: 'error', text: 'Unable to save child identifiers' },
          ]);
        })
        .finally(onFinally);
    } else {
      apiPut(`children/${child.id}`, _child, { accessToken })
        .then(afterSaveSuccess)
        .catch((err) => {
          console.error(err);
          setAlerts([
            { type: 'error', text: 'Unable to save child identifiers' },
          ]);
        })
        .finally(onFinally);
    }
  };

  return (
    <Form<Child>
      className="ChildIdentifiersForm usa-form"
      data={child}
      onSubmit={onFormSubmit}
      noValidate
      autoComplete="off"
      hideStatus={errorsHidden}
    >
      {!hideHeader && <h2>Child's identifiers</h2>}
      {showFieldOrFieldset(child, ['sasid', 'uniqueId']) &&
        (child.organization.uniqueIdType === UniqueIdType.SASID ? (
          <div className="mobile-lg:grid-col-12">
            <SasidField />
          </div>
        ) : child.organization.uniqueIdType === UniqueIdType.Other ? (
          <div className="mobile-lg:grid-col-12">
            <UniqueIdField />
          </div>
        ) : (
          <></>
        ))}
      {showFieldOrFieldset(child, [
        'firstName',
        'lastName',
        'middleName',
        'suffix',
      ]) && (
        <>
          <div className="mobile-lg:grid-col-9">
            <FirstNameField />
          </div>
          <div className="mobile-lg:grid-col-9">
            <MiddleNameField />
          </div>
          <div className="display-flex flex-row flex-align-end grid-row grid-gap">
            <div className="mobile-lg:grid-col-9">
              <LastNameField />
            </div>
            <div className="mobile-lg:grid-col-3">
              <SuffixField />
            </div>
          </div>
        </>
      )}
      {showFieldOrFieldset(child, ['birthdate']) && <DateOfBirthField />}
      {showFieldOrFieldset(child, [
        'birthCertificateType',
        'birthCertificateId',
        'birthTown',
        'birthState',
      ]) && <BirthCertificateFieldSet />}
      <FormSubmitButton
        text={saving ? 'Saving...' : 'Save'}
        disabled={saving}
      />
    </Form>
  );
};
