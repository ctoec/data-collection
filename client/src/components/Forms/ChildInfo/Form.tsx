import React, { useContext, useState } from 'react';
import {
  RaceField,
  EthnicityField,
  GenderField,
  DisabilityServices,
  DualLanguageLearner,
  Foster,
} from './Fields';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { RecordFormProps } from '../types';
import AuthenticationContext from '../../../contexts/AuthenticationContext/AuthenticationContext';
import { Child, RACE_FIELDS } from '../../../shared/models';
import { apiPut } from '../../../utils/api';
import useIsMounted from '../../../hooks/useIsMounted';
import { useValidationErrors } from '../../../hooks/useValidationErrors';
import { getValidationStatusForFields } from '../../../utils/getValidationStatus';

// The fields we use to check to see if this form has errors or missing info
const specialCircumstancesFields = [
  'receivingDisabilityServices',
  'dualLanguageLearner',
  'foster',
];
export const childInfoFields = [
  'americanIndianOrAlaskaNative',
  'asian',
  'blackOrAfricanAmerican',
  'nativeHawaiianOrPacificIslander',
  'white',
  'raceNotDisclosed',
  'hispanicOrLatinxEthnicity',
  'gender',
  ...specialCircumstancesFields,
];
export const doesChildInfoFormHaveErrors = (child?: Child) =>
  child ? !!getValidationStatusForFields(child, childInfoFields) : true;

export const ChildInfoForm = ({
  child,
  afterSaveSuccess,
  hideHeader = false,
  hideErrorsOnFirstLoad = false,
  showFieldOrFieldset = () => true,
  setAlerts,
}: RecordFormProps) => {
  const { accessToken } = useContext(AuthenticationContext);
  const isMounted = useIsMounted();
  const [saving, setSaving] = useState(false);

  if (!child) {
    throw new Error('Child info rendered without child');
  }

  console.log(child)

  const { errorsHidden, setErrorsHidden } = useValidationErrors(
    hideErrorsOnFirstLoad
  );

  const onFinally = () => {
    if (isMounted()) {
      setErrorsHidden(false);
      setSaving(false);
    }
  };

  const onFormSubmit = (_child: Child) => {
    setSaving(true);
    apiPut(`children/${child.id}`, _child, { accessToken })
      .then(afterSaveSuccess)
      .catch((err) => {
        console.error(err);
        setAlerts([
          {
            type: 'error',
            text: 'Unable to save child info',
          },
        ]);
      })
      .finally(onFinally);
  };

  return (
    <Form<Child>
      className="ChildInfoForm usa-form"
      data={child}
      onSubmit={onFormSubmit}
      noValidate
      autoComplete="off"
      hideStatus={errorsHidden}
    >
      {!hideHeader && <h2>Child info</h2>}
      {showFieldOrFieldset(child, [
        ...RACE_FIELDS,
        'hispanixOrLatinxEthnicity',
      ]) && (
          <>
            <RaceField />
            <EthnicityField />
          </>
        )}
      {showFieldOrFieldset(child, ['gender']) && <GenderField />}
      <br />

      {showFieldOrFieldset(child, specialCircumstancesFields) && (
        <h3 className="margin-bottom-0">Special circumstances</h3>
      )}
      {showFieldOrFieldset(child, ['receivesDisabilityServices']) && (
        <DisabilityServices />
      )}
      {showFieldOrFieldset(child, ['dualLanguageLearner']) && (
        <DualLanguageLearner />
      )}
      {showFieldOrFieldset(child, ['foster']) && <Foster />}
      <FormSubmitButton
        text={saving ? 'Saving...' : 'Save'}
        disabled={saving}
      />
    </Form>
  );
};
