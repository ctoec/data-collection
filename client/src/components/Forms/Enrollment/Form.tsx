import React, { useContext, useState } from 'react';
import { Alert, Form, FormSubmitButton } from '@ctoec/component-library';
import { Enrollment, Child } from '../../../shared/models';
import {
  SiteField,
  EnrollmentStartDateField,
  CareModelField,
  AgeGroupField,
  NewFundingField,
} from './Fields';
import UserContext from '../../../contexts/UserContext/UserContext';
import { RecordFormProps } from '../types';
import { getValidationStatusForFields } from '../../../utils/getValidationStatus';
import AuthenticationContext from '../../../contexts/AuthenticationContext/AuthenticationContext';
import { getCurrentEnrollment } from '../../../utils/models';
import { apiPost, apiPut } from '../../../utils/api';
import useIsMounted from '../../../hooks/useIsMounted';
import { useValidationErrors } from '../../../hooks/useValidationErrors';

const enrollmentFields = ['site', 'ageGroup', 'entry', 'model'];
const enrollmentFundingFields = [...enrollmentFields, 'fundings'];
export const doesEnrollmentFormHaveErrors = (
  child?: Child,
  enrollmentId?: number,
  opts: { excludeFundings?: boolean } = {}
) => {
  if (enrollmentId) {
    const enrollment = child?.enrollments?.find((e) => e.id === enrollmentId);
    if (enrollment && !enrollment.fundings?.length) return true;
    return enrollment
      ? !!getValidationStatusForFields(
        enrollment,
        opts.excludeFundings ? enrollmentFields : enrollmentFundingFields
      )
      : false;
  }

  if (child && !!getValidationStatusForFields(child, ['enrollments'])) {
    return true;
  }

  return child?.enrollments?.length
    ? !!getValidationStatusForFields(child.enrollments, enrollmentFundingFields)
    : true;
};

type EnrollmentFormProps = {
  id?: string;
  enrollmentId?: number;
} & RecordFormProps;

export const EnrollmentForm: React.FC<EnrollmentFormProps> = ({
  id,
  child,
  enrollmentId,
  AdditionalButton,
  afterSaveSuccess,
  setAlerts,
  showFieldOrFieldset = () => true,
  hideErrorsOnFirstLoad,
}) => {
  if (!child) {
    throw new Error('Enrollment form rendered without child');
  }

  const { errorsHidden, setErrorsHidden } = useValidationErrors(
    hideErrorsOnFirstLoad
  );

  const [loading, setLoading] = useState(false);
  const { accessToken } = useContext(AuthenticationContext);

  const { user } = useContext(UserContext);
  const sites = user?.sites || [];

  const enrollment =
    (enrollmentId
      ? child.enrollments?.find((e) => e.id === enrollmentId)
      : getCurrentEnrollment(child)) || ({} as Enrollment);

  const createEnrollment = async (updatedData: Enrollment) =>
    apiPost(
      `children/${child.id}/change-enrollment`,
      { newEnrollment: updatedData },
      { accessToken, jsonParse: false }
    );

  const updateEnrollment = async (updatedData: Enrollment) =>
    apiPut(`enrollments/${enrollment.id}`, updatedData, {
      accessToken,
    });

  const saveData = enrollment.id ? updateEnrollment : createEnrollment;

  const isMounted = useIsMounted();

  const onFinally = () => {
    if (isMounted()) {
      setErrorsHidden(false);
      setLoading(false);
    }
  };
  const onSubmit = (updatedData: Enrollment) => {
    setLoading(true);

    //  Because the radio group for the Site field broadcasts ID as a string,
    //  but the API is expecting a number (as it should)
    if (
      !!updatedData.site &&
      !!updatedData.site.id &&
      typeof updatedData.site.id === 'string'
    ) {
      updatedData.site.id = parseInt(updatedData.site.id);
    }
    saveData(updatedData)
      .then(afterSaveSuccess)
      .catch((err) => {
        console.error(err);
        setAlerts([
          {
            type: 'error',
            text: 'Unable to save enrollment',
          },
        ]);
      })
      .finally(onFinally);
  };

  return (
    <>
      {!enrollment.id && (
        <Alert
          type="info"
          text="Adding a child who has withdrawn, changed sites or age groups? Add their earliest enrollment then make changes from the child's record in your roster."
        />
      )}
      <Form<Enrollment>
        id={id || `edit-enrollment-form-${enrollment.id}`}
        data={enrollment}
        onSubmit={onSubmit}
        hideStatus={errorsHidden}
      >
        {showFieldOrFieldset(enrollment, ['site']) && (
          <SiteField<Enrollment>
            sites={sites.filter(
              (s) => s.organizationId === child?.organization?.id
            )}
          />
        )}
        {showFieldOrFieldset(enrollment, ['entry', 'model']) && (
          <>
            <EnrollmentStartDateField<Enrollment> />
            <CareModelField<Enrollment> />
          </>
        )}
        {showFieldOrFieldset(enrollment, ['ageGroup']) && (
          <AgeGroupField<Enrollment> />
        )}
        {showFieldOrFieldset(enrollment, ['fundings']) && (
          <NewFundingField<Enrollment>
            fundingAccessor={(data) => data.at('fundings').at(0)}
            getEnrollment={(data) => data.value}
            organizationId={child?.organization?.id}
          />
        )}
        {AdditionalButton}
        <FormSubmitButton
          text={loading ? 'Saving...' : 'Save'}
          disabled={loading}
        />
      </Form>
    </>
  );
};
