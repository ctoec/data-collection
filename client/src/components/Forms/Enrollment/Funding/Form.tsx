import React, { useState, useContext } from 'react';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { Funding, Child } from '../../../../shared/models';
import { ContractSpaceField, ReportingPeriodField } from './Fields';
import { RecordFormProps } from '../../types';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';
import { apiPut } from '../../../../utils/api';
import {
  getCurrentEnrollment,
  getCurrentFunding,
} from '../../../../utils/models';
import { getValidationStatusForFields } from '../../../../utils/getValidationStatus';
import { NewFundingField } from '../Fields';
import { useValidationErrors } from '../../../../hooks/useValidationErrors';

const fundingFields = [
  'fundingSpace',
  'firstReportingPeriod',
  'lastReportingPeriod',
];
export const doesFundingFormHaveErrors = (
  child?: Child,
  enrollmentId?: number,
  fundingId?: number
) => {
  if (enrollmentId && fundingId) {
    const funding = child?.enrollments
      ?.find((enrollment) => enrollment.id === enrollmentId)
      ?.fundings?.find((funding) => funding.id === fundingId);
    return funding
      ? !!getValidationStatusForFields(funding, fundingFields)
      : false;
  }
  return false;
};

type FundingFormProps = {
  id: string;
  enrollmentId: number;
  fundingId: number;
} & RecordFormProps;

export const FundingForm: React.FC<FundingFormProps> = ({
  id,
  child,
  enrollmentId,
  fundingId,
  AdditionalButton,
  setAlerts,
  afterSaveSuccess,
  hideErrorsOnFirstLoad,
}) => {
  if (!child) {
    throw new Error('Funding form rendered without child');
  }

  const enrollment = enrollmentId
    ? child.enrollments?.find((e) => e.id === enrollmentId)
    : getCurrentEnrollment(child);

  if (!enrollment) {
    throw new Error('Funding form rendered without enrollment');
  }

  // TODO: is an additional funding still created when a user enters incomplete info in create child, then enters different info and saves?
  // If so it's maybe bc we're not using the right funding here to fill in the info the second time around/not updating the right funding
  const funding = fundingId
    ? enrollment?.fundings?.find((f) => f.id === fundingId)
    : getCurrentFunding({ enrollment });

  if (!funding) {
    throw new Error('Funding form rendered without funding');
  }

  const { errorsHidden, setErrorsHidden } = useValidationErrors(
    hideErrorsOnFirstLoad
  );

  const { accessToken } = useContext(AuthenticationContext);
  const [loading, setLoading] = useState(false);
  const onSubmit = (updatedData: Funding) => {
    setLoading(true);
    apiPut(`/enrollments/${enrollmentId}/fundings/${fundingId}`, updatedData, {
      accessToken,
    })
      .then(afterSaveSuccess)
      .catch((err) => {
        console.error(err);
        setAlerts([
          {
            type: 'error',
            text: 'Unable to save funding',
          },
        ]);
      })
      .finally(() => {
        setLoading(false);
        setErrorsHidden(false);
      });
  };

  // If the funding has a funding space, render
  // the normal edit funding form (where only contract space,
  // first and last reporting periods are editable).
  // If the funding does NOT have a funding space, render
  // a form containing the "new funding" field, to enable
  // the user to select a funding source, contract space
  // and reporting periods
  return funding.fundingSpace ? (
    <Form<Funding>
      id={id}
      className="usa-form"
      data={funding}
      onSubmit={onSubmit}
      hideStatus={errorsHidden}
    >
      <ContractSpaceField<Funding>
        ageGroup={enrollment.ageGroup}
        fundingSource={funding.fundingSpace.source}
        organizationId={child.organization.id}
      />
      <ReportingPeriodField<Funding>
        fundingSource={funding.fundingSpace.source}
        accessor={(data) => data.at('firstReportingPeriod')}
        showStatus
      />
      {/* Only display last reporting period field if a value already exists */}
      {!!funding.lastReportingPeriod && (
        <ReportingPeriodField<Funding>
          fundingSource={funding.fundingSpace.source}
          accessor={(data) => data.at('lastReportingPeriod')}
          isLast={true}
          showStatus
        />
      )}
      {AdditionalButton}
      <FormSubmitButton
        text={loading ? 'Saving...' : 'Save'}
        disabled={loading}
      />
    </Form>
  ) : (
    <Form<Funding>
      id={id}
      className="usa-form"
      data={funding}
      onSubmit={onSubmit}
      hideStatus={errorsHidden}
    >
      <NewFundingField<Funding>
        getEnrollment={() => enrollment}
        organizationId={child.organization.id}
        isEdit={true}
      />
      {AdditionalButton}
      <FormSubmitButton
        text={loading ? 'Saving...' : 'Save'}
        disabled={loading}
      />
    </Form>
  );
};
