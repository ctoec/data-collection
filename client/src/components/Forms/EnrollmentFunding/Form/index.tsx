import React, { useEffect } from 'react';
import { ChangeEnrollmentForm } from './ChangeEnrollmentForm';
import { EditEnrollmentForm } from './EditEnrollmentForm';
import { EditFundingForm } from './EditFundingForm';
import { ChangeFundingForm } from './ChangeFundingForm';
import { EditFormProps } from '../../types';
import { useSites } from '../../../../hooks/useSites';
import { useFundingSpaces } from '../../../../hooks/useFundingSpaces';
import { useReportingPeriods } from '../../../../hooks/useReportingPeriods';

// The fields we use to check to see if this form has errors or missing info
export const enrollmentFundingFields = {
  enrollments: [],
};

export const EnrollmentFundingForm: React.FC<EditFormProps> = ({
  child,
  onSuccess,
  setAlerts,
  reportingPeriods: inputReportingPeriods,
}) => {
  // Get site options for new enrollments
  const { sites } = useSites();

  // Get fundingSpaces for new fundings
  const { fundingSpaces } = useFundingSpaces();

  // Get reporting periods (needed to update enrollments with fundings)
  const { reportingPeriods } = useReportingPeriods(inputReportingPeriods);

  // Clear any previously displayed alerts from other tabs
  useEffect(() => {
    setAlerts([]);
  }, []);

  // Separate enrollments into current (no end date) and past
  // (with end date). Either may not exist
  if (!child) {
    return <></>;
  }

  const enrollments = child.enrollments || [];
  const currentEnrollment = enrollments.find((e) => !e.exit);
  const pastEnrollments = currentEnrollment
    ? enrollments.filter((e) => e.id !== currentEnrollment.id)
    : enrollments;

  return (
    <>
      <h2>Enrollment and funding</h2>
      <ChangeEnrollmentForm
        reportingPeriods={reportingPeriods}
        fundingSpaces={fundingSpaces}
        childName={child.firstName || ''}
        currentEnrollment={currentEnrollment}
        childId={child.id}
        sites={sites}
        onSuccess={onSuccess}
      />
      {currentEnrollment && (
        <>
          <h3>Current enrollment</h3>
          <EditEnrollmentForm
            key="edit-current-enrollment"
            isCurrent={true}
            enrollment={currentEnrollment}
            onSuccess={onSuccess}
          />
          {currentEnrollment.fundings?.map((funding) => (
            <EditFundingForm
              key={funding.id}
              isCurrent={true}
              reportingPeriods={reportingPeriods}
              fundingSpaces={fundingSpaces}
              funding={funding}
              enrollment={currentEnrollment}
              onSuccess={onSuccess}
            />
          ))}
          <ChangeFundingForm
            fundingSpaces={fundingSpaces}
            reportingPeriods={reportingPeriods}
            enrollment={currentEnrollment}
            onSuccess={onSuccess}
          />
        </>
      )}
      {!!pastEnrollments.length && (
        <>
          <h3>Past enrollments</h3>
          {pastEnrollments.map((enrollment) => (
            <>
              <EditEnrollmentForm
                key={enrollment.id}
                enrollment={enrollment}
                onSuccess={onSuccess}
              />
              {enrollment.fundings?.map((funding) => (
                <EditFundingForm
                  key={funding.id}
                  reportingPeriods={reportingPeriods}
                  fundingSpaces={fundingSpaces}
                  funding={funding}
                  enrollment={enrollment}
                  onSuccess={onSuccess}
                />
              ))}
            </>
          ))}
        </>
      )}
    </>
  );
};
