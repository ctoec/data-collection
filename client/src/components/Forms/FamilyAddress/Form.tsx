import React, { useState, useContext } from 'react';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { Child, Family } from '../../../shared/models';
import { AddressFieldset, HomelessnessField } from './Fields';
import AuthenticationContext from '../../../contexts/AuthenticationContext/AuthenticationContext';
import { apiPut } from '../../../utils/api';
import { RecordFormProps } from '../types';
import useIsMounted from '../../../hooks/useIsMounted';
import { useValidationErrors } from '../../../hooks/useValidationErrors';
import { getValidationStatusForFields } from '../../../utils/getValidationStatus';

const familyAddressFields = [
  'streetAddress',
  'town',
  'state',
  'zipCode',
  'homelessness',
];
export const doesFamilyAddressFormHaveErrors = (child?: Child) =>
  child?.family
    ? !!getValidationStatusForFields(child.family, familyAddressFields)
    : true;

/*
 * Functional component that allows a user to modify the address
 * information of a family nested within a Child object. The
 * family data is manipulated directly to avoid interactions
 * with the Child object that the family form doesn't need
 * to know about.
 */
export const FamilyAddressForm: React.FC<RecordFormProps> = ({
  record: child,
  afterDataSave,
  hideHeader = false,
  hideErrorsOnFirstLoad = false,
}) => {
  const { accessToken } = useContext(AuthenticationContext);
  const isMounted = useIsMounted();
  const [saving, setSaving] = useState(false);

  if (!child) {
    throw new Error('Family info rendered without child');
  }

  const { family: inputFamily = {} as Family } = child;

  const { obj: family, setErrorsHidden } = useValidationErrors<Family>(
    inputFamily,
    hideErrorsOnFirstLoad
  );

  const onSubmit = (newState: Family) => {
    setErrorsHidden(false);
    setSaving(true);
    apiPut(`families/${family.id}`, newState, { accessToken })
      .then(afterDataSave)
      .catch((err) => {
        console.log(err);
      })
      .finally(() => (isMounted() ? setSaving(false) : null));
  };

  return (
    <div className="grid-container margin-top-2">
      {!hideHeader && <h2 className="grid-row">Family Address</h2>}
      <Form<Family>
        className="FamilyAddressForm"
        data={family}
        onSubmit={onSubmit}
        noValidate
        autoComplete="off"
      >
        <AddressFieldset />
        <HomelessnessField />
        <div className="grid-row margin-top-2">
          <FormSubmitButton
            text={saving ? 'Saving...' : 'Save edits'}
            disabled={saving}
          />
        </div>
      </Form>
    </div>
  );
};
