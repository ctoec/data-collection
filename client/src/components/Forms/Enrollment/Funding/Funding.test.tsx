import React from 'react';
import {
  snapshotTestHelper,
  accessibilityTestHelper,
} from '../../../../testHelpers';
import { FundingForm } from './Form';
import {
  Family,
  Organization,
  Child,
  Site,
  FundingSource,
  FundingTime,
  Funding,
} from '../../../../shared/models';
import moment from 'moment';

const child = {
  name: 'First Last',
  firstName: 'First',
  lastName: 'Last',
  id: '00000000-0000-0000-0000-000000000000',
  family: {} as Family,
  organization: { id: 1 } as Organization,
  enrollments: [
    {
      id: 1,
      child: {} as Child,
      childId: '',
      site: { siteName: 'Site A', organization: { id: 1 } } as Site,
      entry: moment.utc('2020-09-03'),
    },
    {
      id: 2,
      child: {} as Child,
      childId: '',
      site: { siteName: 'Site B', organization: { id: 1 } } as Site,
      entry: moment.utc('2019-09-03'),
      exit: moment.utc('2020-08-01'),
      fundings: [
        {
          id: 1,
          fundingSpace: {
            source: FundingSource.CDC,
            time: FundingTime.FullTime,
          },
          firstReportingPeriod: { period: moment.utc('2019-09-01') },
          lastReportingPeriod: { period: moment.utc('2020-07-01') },
        } as Funding,
      ],
    },
  ],
};

describe('Forms', () => {
  describe('Funding', () => {
    snapshotTestHelper(
      <FundingForm
        id={'id'}
        child={child}
        enrollmentId={2}
        fundingId={1}
        afterSaveSuccess={jest.fn()}
        setAlerts={jest.fn()}
        topHeadingLevel="h2"
      />,
      { wrapInRouter: true }
    );

    accessibilityTestHelper(
      <FundingForm
        id={'id'}
        child={child}
        enrollmentId={2}
        fundingId={1}
        afterSaveSuccess={jest.fn()}
        setAlerts={jest.fn()}
        topHeadingLevel="h2"
      />,
      { wrapInRouter: true }
    );
  });
});
