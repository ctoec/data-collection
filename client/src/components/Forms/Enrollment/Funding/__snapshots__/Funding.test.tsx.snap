// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forms Funding matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="usa-form usa-form"
    id="id"
  >
    <div
      class="usa-form-group"
    >
      <label
        class="usa-label"
        for="contract-space"
      >
        Contract space
      </label>
      <select
        aria-required="true"
        class="usa-select"
        id="contract-space"
        name="contract-space"
      >
        <option
          value=""
        >
          - Select -
        </option>
      </select>
    </div>
    <fieldset
      aria-labelledby="fieldset-legend-funding-start-date funding-start-date-hint"
      class="usa-fieldset"
      id="funding-start-date"
    >
      <legend
        id="fieldset-legend-funding-start-date"
      >
        <span
          class="usa-label"
        >
          Funding start date
        </span>
      </legend>
      <span
        class="usa-hint text-italic"
        id="funding-start-date-hint"
      >
        For example: 06/21/2008
      </span>
      <div>
        <div
          class="bx--form-item"
        >
          <div
            class="bx--date-picker bx--date-picker--single"
            id="funding-start-date-picker"
          >
            <div
              class="bx--date-picker-container"
            >
              <label
                class="bx--label bx--visually-hidden"
                for="funding-start-date-input"
              >
                Date picker
              </label>
              <div
                class="bx--date-picker-input__wrapper"
              >
                <input
                  class="bx--date-picker__input flatpickr-input"
                  id="funding-start-date-input"
                  pattern="\\\\d{1,2}\\\\/\\\\d{1,2}\\\\/\\\\d{4}"
                  placeholder="mm/dd/yyyy"
                  type="text"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="bx--date-picker__icon"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M26,4h-4V2h-2v2h-8V2h-2v2H6C4.9,4,4,4.9,4,6v20c0,1.1,0.9,2,2,2h20c1.1,0,2-0.9,2-2V6C28,4.9,27.1,4,26,4z M26,26H6V12h20	V26z M26,10H6V6h4v2h2V6h8v2h2V6h4V10z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
    <fieldset
      aria-labelledby="fieldset-legend-funding-end-date funding-end-date-hint"
      class="usa-fieldset"
      id="funding-end-date"
    >
      <legend
        id="fieldset-legend-funding-end-date"
      >
        <span
          class="usa-label"
        >
          Funding end date
        </span>
      </legend>
      <span
        class="usa-hint text-italic"
        id="funding-end-date-hint"
      >
        For example: 06/21/2008
      </span>
      <div>
        <div
          class="bx--form-item"
        >
          <div
            class="bx--date-picker bx--date-picker--single"
            id="funding-end-date-picker"
          >
            <div
              class="bx--date-picker-container"
            >
              <label
                class="bx--label bx--visually-hidden"
                for="funding-end-date-input"
              >
                Date picker
              </label>
              <div
                class="bx--date-picker-input__wrapper"
              >
                <input
                  class="bx--date-picker__input flatpickr-input"
                  id="funding-end-date-input"
                  pattern="\\\\d{1,2}\\\\/\\\\d{1,2}\\\\/\\\\d{4}"
                  placeholder="mm/dd/yyyy"
                  type="text"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="bx--date-picker__icon"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M26,4h-4V2h-2v2h-8V2h-2v2H6C4.9,4,4,4.9,4,6v20c0,1.1,0.9,2,2,2h20c1.1,0,2-0.9,2-2V6C28,4.9,27.1,4,26,4z M26,26H6V12h20	V26z M26,10H6V6h4v2h2V6h8v2h2V6h4V10z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
    <button
      class="usa-button bx--btn bx--btn--primary"
      tabindex="0"
      type="submit"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
